### PR TITLE
[SPARK-14491][SQL] refactor object operator framework to make it easy to eliminate serializations

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This PR tries to separate the serialization and deserialization logic from object operators, so that it's easier to eliminate unnecessary serializations in optimizer.

Typed aggregate related operators are special, they will deserialize the input row to multiple objects and it's difficult to simply use a deserializer operator to abstract it, so we still mix the deserialization logic there.
## How was this patch tested?

existing tests and new test in `EliminateSerializationSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @davies @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2236,16 +2236,16 @@ class Dataset[T] private[sql](
...
+   * Represents the content of the [[Dataset]] as an [[RDD]] of [[T]].
    *
    * @group rdd
    * @since 1.6.0
    */
   lazy val rdd: RDD[T] = {
-    queryExecution.toRdd.mapPartitions { rows =>
-      rows.map(boundTEncoder.fromRow)
+    val objectType = unresolvedTEncoder.deserializer.dataType
+    val deserialized = CatalystSerde.deserialize[T](logicalPlan)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Now we have chance to save the final deserialization.

**Comment by yhuai** [comment]
@davies Can you also review this? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 2]
```diff
@@ -474,6 +474,10 @@ case class CollapseCodegenStages(conf: SQLConf) extends Rule[SparkPlan] {
    * Inserts a WholeStageCodegen on top of those that support codegen.
    */
   private def insertWholeStageCodegen(plan: SparkPlan): SparkPlan = plan match {
+    // For operators that will output domain object, do not insert WholeStageCodegen for it as
+    // domain object can not be written into unsafe row.
+    case plan if plan.output.length == 1 && plan.output.head.dataType.isInstanceOf[ObjectType] =>
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: We could update WholeStageCodegen to support GenericInternalRow as well (could be another PR).

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master, thanks!
